ouroboros-consensus-test/src/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -0,0 +1,638 @@
+    --
+    -- 2) The stationary distribution is the eigenvector with eigenvalue 1.
+    --
+    -- 3) We want the /left/ eigenvector (TODO is this fundamental to transition matrices?).
+    --
+    -- Thus we want a vector v such that
+    --
+    -- > v A       = v
+    -- > v A  - v  = 0
+    -- > v (A - I) = 0
+    --
+    -- We find that using 'Mat.null1' of the transpose of A - I.
+    sd :: Mat.Vector Double
+    sd = Mat.null1 $ Mat.tr $ p - Mat.ident sz
+
+{-
+    This would be equally (?) exact, but it makes the test suite about 2-3
+    times slower than does 'Mat.null1'.
+
+    sd :: Mat.Vector Double
+    sd = lvecs Mat.! Mat.maxIndex vals
+
+    (Mat.cmap realPart -> vals, Mat.tr -> Mat.cmap realPart -> lvecs) = Mat.eig (Mat.tr p)
+-}
+
+{-
+    TODO This makes the test suite 3-5 times faster than does 'Mat.null1', but
+    I don't know how to bound its inaccuracy. Maybe there's an argument that
+    for some X after X * (s-d) steps (instead of 1024) it's certainly "close
+    enough" to the stationary distribution for our specific use case?
+
+    -- a brute-force way to (hopefully) find the stationary distribution
+    -- (recall the way in which 'stimes' is optimized) sd :: Mat.Vector Double
+    -- sd = (Mat.! 0) $ stimes (2^(10 :: Int) :: Int) p
+-}
+
+    -- Transition matrix of an Markov Chain where each state is the
+    -- (idealized?) number of active slots in an EHCG window.
+    --
+    -- (This is a /stochastic matrix/ because every /row/ sums to 1. And every
+    -- state is reachable from every other, since @0 < asc@.)
+    --
+    -- We have one lingering doubt about whether this system perfectly models
+    -- the behavior of the loop in 'uniformTheHonestChain' that visits the
+    -- remaining windows, hence the "idealized?" above. In particular, these
+    -- transitions are the result of assuming that the active slots within a
+    -- window are uniformly distributed after it has been visited by that loop.
+    -- Specifically, we assume that the probability of the oldest slot being
+    -- active is the population count divided by the window width. Our
+    -- lingering doubt is that we're overestimating the degrees of freedom such
+    -- that somehow a slot's value from /before/ some window is unexpectedly
+    -- influencing the value of the first slot in that window in a way that is
+    -- not accounted for by the number of active slots in the preceding window,
+    -- ie that the Markov Property does not actually hold. I think the seed for
+    -- this doubt was Esgen's earlier insight that a chain with /exactly/ @n@
+    -- actives in every @d@ slot window is periodic. In that case, if there are
+    -- more than @d@ contiguous windows that each have exactly @n@ actives,
+    -- then there is perfect auto-correlation throughout those windows, and it
+    -- spans farther than one window. That's our suspicion for the kind of
+    -- counterexample of the Markov Property of this transition matrix.
```

Comment:
This should be reworded; any stochastic matrix (such as the one below) corresponds to a Markov chain, but the question is whether that actually models what we are interested in (i.e. whether it is "memoryless").

Reply:
Fixed 👍